### PR TITLE
feature(gripper): add precise linear motion control for brushed motors

### PR DESCRIPTION
**Pull request by ryanthecoder**
Will need to use RET-1097-gripper-evt-basic-position-control branch in the mono-repo to send the new can messages

### CODE REVIEW

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (fsinapi) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:80]: I'm pretty sure that calculating the PID output from an absolute value results in some fundamental changes to how the controller functions. In particular, the integral term will _only_ ever become more positive because every error value is positive. I would generally expect a system like this to calculate the correction value from a signed error value, and then apply the positive/negative direction and convert to an unsigned PWM value after the fact.

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:61]:
```suggestion
    // upon advice from hardware, 0.01mm is a good limit for precision
```

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:168]: why would this be an error rather than an immediate completion?

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:74]: let's either always use `grip`/`ungrip` or never use `grip`/`ungrip`

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:63]: this is the only place we ever use this; can we make `acceptable_position_error` be a value parameter to this class's ctor instead and save a float calculation?

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:97]: this is getting a little complex; maybe we should split it into like an `execute_active_move()` subfunction and a `execute_idle()` subfunction?

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:124]: the `if( (...) abs(pulses() - position) < acceptable)` is the first thing `controlled_move_to` does, right? can we cut it from this `if` and rely on the function call?

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:227]: does this need to be atomic? it's only accessed inside the ISR, right?

Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:227]: also, does it make more sense for this to be an enum of something like
```
enum class State {
   FORCE_CONTROLLING,
   POSITION_CONTROLLING,
   IDLE 
};
```
rather than a `holding` bool and an implicit "don't disable the motor when you finish a force control move"?

File: include/can/core/messages.hpp
Comment (sfoster1) [inline include/can/core/messages.hpp:727]: why do we need `duration` and `duty_cycle`? those values aren't used, right?

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Looks really good overall, would love a demo. Some details I think we could improve though.

File: include/can/core/messages.hpp
Comment (ryanthecoder) [reply include/can/core/messages.hpp:727]: duration currently isn't used by any of the gripper moves, we have a ticket in our queue to use duration to trigger a timeout so moves don't get stuck forever.
and on the monorepo side all the gripper messages have the same payload type for consistency there, so duty cycle is a hold over, I may use this in the future as a duty_cycle MAX value

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (ahiuchingau) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:31]: Oh you're absolutely right about this. I forgot to factor in the period!

We probably want to shorten this hold-off period a bit.

File: gripper/firmware/interfaces_grip_motor.cpp
Comment (ahiuchingau) [inline gripper/firmware/interfaces_grip_motor.cpp:46]: Could you add the formula for getting these pid values to the comment here?

**Review by ahiuchingau — APPROVED** [review]
Very cool! Can't wait to see the behavior in real life.

File: include/motor-control/firmware/brushed_motor/brushed_motor_hardware.hpp
Comment (sfoster1) [inline include/motor-control/firmware/brushed_motor/brushed_motor_hardware.hpp:40]:
```suggestion
                          1.f / config.encoder_interrupt_freq} {}
```

just to make it explicit

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Sounds good to me on the message elements front. Couple more little things but this is looking really good.

File: gripper/firmware/interfaces_grip_motor.cpp
Comment (Laura-Danielle) [inline gripper/firmware/interfaces_grip_motor.cpp:49]: Is this an incomplete thought?

File: gripper/firmware/motor_encoder_hardware.h
Comment (Laura-Danielle) [inline gripper/firmware/motor_encoder_hardware.h:15]: Should we make a define for the encoder interrupt frequency as well -- since we're already doing one for the gripper timer frequency?

File: include/can/core/messages.hpp
Comment (Laura-Danielle) [inline include/can/core/messages.hpp:725]: What's the difference between this and a "grip" request? Can't we just combine the two CAN messages since we're really just enhancing the functionality of gripping?

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (Laura-Danielle) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:60]: Should this also be absolute value since we're comparing it below to the abs of the `move_delta`?

**Review by Laura-Danielle — COMMENTED** [review]
Some small comments that definitely don't block merging this PR.

File: gripper/firmware/interfaces_grip_motor.cpp
Comment (sfoster1) [reply gripper/firmware/interfaces_grip_motor.cpp:49]: definitely want to fix this before merging

File: include/can/core/messages.hpp
Comment (sfoster1) [reply include/can/core/messages.hpp:725]: the grip request specifies force control, and this specifies position control. they carry similar bits for now, but that won't necessarily always be true - they mean different things

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:39]:
```suggestion
static constexpr double ACCEPTABLE_DISTANCE_TOLERANCE_MM = 0.01;
```

File: include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp
Comment (sfoster1) [inline include/motor-control/core/brushed_motor/brushed_motor_interrupt_handler.hpp:61]:
```suggestion
                    ACCEPTABLE_DISTANCE_TOLERANCE_MM);
```

**Review by sfoster1 — APPROVED** [review]
This is real nice! IMO we should fix that comment @Laura-Danielle noted and the typo in tolerance but looks good!

**Pull request closed without merging**
